[Lim Jun Hong, Don] iP

donljh opened the pull request:
# Blob
> “We’re all victims of victims. How far back do you want to go, searching for the source? It’s better to start with yourself.” - Edith Eger ([source](https://www.goodreads.com/work/quotes/74535739-the-gift-12-lessons-to-save-your-life))
Use Blob to help you remember your tasks, deadlines and events! Blob is:
- Text-based
- Easy to learn
- BLAZINGLY fast to use

To get started:
1. Download it from [here](https://github.com/donljh/ip/releases/download/A-Jar/blob.jar)
2. Double-click to open it
3. Tell Blob the tasks you want Blob to remember

All for **FREE**!

Features:

- [x] Managing tasks, deadlines and events 📝 
- [ ] Cooking your meals (coming soon) 🍳
- [ ] ~~World domination (work in progress) 🌏~~

Change your saved tasks file in the `main` method

```java
public static void main(String[] args) {
    Blob blob = new Blob("data/tasks.txt");
    blob.start();
}
```

Commit:
Add Gradle support

donljh committed:
Add greeting and command echoing

donljh committed:
Level 2: Add, List

donljh committed:
Level 3: Mark as Done

donljh committed:
Add task type classes

donljh committed:
Level 4: ToDos, Events, Deadlines

donljh committed:
A-TextUiTesting

donljh committed:
Add exception classes

donljh committed:
Level 5: Handle Errors

donljh committed:
Level 6: Delete

donljh committed:
A-Enums: Use Enum

donljh committed:
Ammend test scripts

donljh committed:
Level 7: Add task saving and loading

donljh committed:
Level 8: Add date parsing functionality

donljh committed:
Merge branch 'branch-Level-7'

donljh committed:
Merge branch 'branch-Level-8'

donljh committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

donljh committed:
Add Ui package, TextUi class, extract welcome and goodbye messages to Messages class

donljh committed:
Refactor main application to incorporate TextUi class

donljh committed:
Add Command class, CommandType enum and respective subclass skeletons

donljh committed:
Add TaskList, TaskCommand classes

donljh committed:
Add DateTimeParser class, move command logic to Command classes

donljh committed:
Add DateTimeParser, Parser class and move command logic into subclasses

donljh committed:
Repackage files

donljh committed:
Fix indentation errors in print statements, amend exceptions

donljh committed:
Add Storage class

donljh committed:
Add unit tests

donljh committed:
Add more JavaDocs

donljh committed:
Fix code styling and imports

donljh committed:
Add find functionality

donljh committed:
Include missing JavaDoc

donljh committed:
Merge branch 'branch-A-JavaDoc'

donljh committed:
Merge branch 'branch-Level-9'

donljh committed:
Fix minor bugs

seanmanik commented on src/main/java/blob/common/Messages.java:
```diff
@@ -0,0 +1,36 @@
+/**
+ * Class containing the reusable messages to print to user
+ */
+public class Messages {
```
I like how messages are all stored in a single Messages class. I think this is a very creative idea!

seanmanik commented on src/main/java/blob/common/Messages.java:
```diff
@@ -0,0 +1,36 @@
+ * Class containing the reusable messages to print to user
+ */
+public class Messages {
+    public static final String MESSAGE_WELCOME_1 = "Hello... me Blob...";
```
These variables are very clearly named and is descriptive, which helps readers to better understand your code.

seanmanik commented on src/main/java/blob/Blob.java:
```diff
@@ -0,0 +1,78 @@
+    /**
+     * Start the interaction with an instance of blob.Blob
+     */
+    public void start() {
```
The code here is really clean 👍

seanmanik commented on src/main/java/blob/parser/DateTimeParser.java:
```diff
@@ -0,0 +1,58 @@
+
+import java.util.Locale;
+
+public class DateTimeParser {
```
Creating a DateTimeParser class really helps out a lot in parsing the different types of date formats available. Great job!

seanmanik commented on src/main/java/blob/tasks/Task.java:
```diff
@@ -0,0 +1,62 @@
+     *
+     * @return A checkmark icon (✓) if the task is done, and " " otherwise
+     */
+    public String getStatusIcon() {
```
Excellent use of a tick instead of a cross as suggested. Creative and user-friendly.

seanmanik commented on src/main/java/blob/tasks/Event.java:
```diff
@@ -0,0 +1,43 @@
+    public Event(String description, String at) throws InvalidDateFormatException {
+        super(description, TaskType.EVENT);
+        try {
+            this.at = new DateTimeParser().parse(at);
```
It is great that this the "at" date of events is also parsed. Would the bot be expecting the same formats for both events and deadlines? If so, this would work well. Otherwise, one suggestion may be to create a separate "at" parser for events.

seanmanik reviewed (commented):
Author has utilised OOP principles in across his code, and has adhered to the provided coding standards. Code was easy to understand and variable names allowed the reader to easily identify exactly what their purpose is. Overall, an excellent submission!

donljh committed:
Configure checkstyle, update code to fit checkstyling rules

donljh committed:
Add GUI

donljh committed:
Merge branch 'branch-Level-10'

richavm14 commented on src/main/java/blob/storage/Storage.java:
```diff
@@ -0,0 +1,93 @@
+     * @return A TaskList based on parsing task data in the specified file path
+     * @throws ErrorLoadingTaskException
+     */
+    public TaskList loadTaskList() throws ErrorLoadingTaskException {
```
This code is readable and avoids deep nesting!

richavm14 commented on src/main/java/blob/tasks/Task.java:
```diff
@@ -0,0 +1,68 @@
+     */
+    public String toFileString() {
+        return String.format("%s | %d | %s", this.type.getAbbreviation(),
+                this.isDone ? 1 : 0, this.description);
```
This code is also readable! (due to line break)

richavm14 commented on src/main/java/blob/ui/TextUi.java:
```diff
@@ -0,0 +1,80 @@
+ */
+public class TextUi {
+    /** Divider to separate message instances by the Blob text UI */
+    private static final String MESSAGE_DIVIDER = "=".repeat(100);
```
I like that you have used named constants to make the code easier to understand!

richavm14 commented on src/main/java/blob/parser/Parser.java:
```diff
@@ -0,0 +1,195 @@
+/**
+ * The Parser class deals with making sense of user commands.
+ */
+public class Parser {
```
Code is structured logically 👍

richavm14 reviewed (commented):
Generally you have maintained good code quality. Good work!

donljh committed:
Fix checkmark display on GUI, change className in build.gradle file

donljh committed:
Add assertions

donljh committed:
Remove unused code

donljh committed:
Add more comments and linebreaks

donljh committed:
Merge pull request #2 from donljh/branch-A-Assertions

Add assertions

donljh committed:
Merge pull request #1 from donljh/branch-A-CodeQuality

Improve Code Quality

donljh committed:
Add missing JavaDoc comments

donljh committed:
Update Task class to include priority attribute

Task classes/objects have no priority attribute, hence it is impossible
to differentiate between tasks of different priorities

Introducing a priority attribute to tasks will allow further priority
related functions to be implemented in the future.

We introduce this attribute into the abstract Task class to make use of
inheritance, so that all inhering task-related classes will have access
to the priority attribute.

donljh committed:
Update Parser class to accept task priorities

Users have no way to tag tasks with priorities upon creation.

Users might want to use priorities to differentiate their tasks on
task creation.

Allowing the Parser class to expect and parse task priorities, users
can input task priorities following a task command in the form
as follows: todo/event/deadline <priority> <task info> ... where
priority is of the format: //h, //m or //l.

Priority takes precedence over task information both in task creation
and listing out of tasks because it is easier for the user to focus on
task priority first then the other task information which are lengthier.

donljh committed:
Fix minor checkstyle errors

donljh committed:
Add User Guide and UI screenshot

donljh committed:
Fix GUI font issues

donljh committed:
Fix build issues
